File: Machine Learning A-Z Template Folder/Part 2 - Regression/Section 5 - Multiple Linear Regression/mlr_rm.py
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
# from sklearn.impute import SimpleImputer
from sklearn.preprocessing import LabelEncoder, OneHotEncoder, StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.compose import ColumnTransformer
import statsmodels.formula.api as sm


def backwardsElimination(x, sl):
    numVars = len(x[0])
    for i in range(0, numVars):
        regressor_OLS = sm.OLS(y, x).fit()
        maxVar = max(regressor_OLS.pvalues).astype(float)
        if maxVar > sl:
            for j in range(0, numVars - i):
                if (regressor_OLS.pvalues[j].astype(float) == maxVar):
                    x = np.delete(x, j, 1)
    print(regressor_OLS.summary())
    return x


data_set = pd.read_csv('/Users/rmaffesoli/Documents/personal/ML_A_Z/Machine Learning A-Z Template Folder/Part 2 - Regression/Section 5 - Multiple Linear Regression/50_Startups.csv')
X = data_set.iloc[:, :-1].values
y = data_set.iloc[:, -1].values

labelencoder_X = LabelEncoder()
X[:, 3] = labelencoder_X.fit_transform(X[:, 3])
onehotencoder = OneHotEncoder(categorical_features=[3])
X = onehotencoder.fit_transform(X).toarray()
X = X[:, 1:]

# transformer = ColumnTransformer(transformers=[('oneHot', OneHotEncoder(), [3])])
# X = transformer.fit_transform(X)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

regressor = LinearRegression()
regressor.fit(X_train, y_train)
y_pred = regressor.predict(X_test)

X = np.append(arr=np.ones((50, 1)).astype(int), values=X, axis=1)
X_opt = X[:, [0, 1, 2, 3, 4, 5]]
SL = 0.05
X_modeled = backwardsElimination(X_opt, SL)

File: Machine Learning A-Z Template Folder/Part 2 - Regression/Section 5 - Multiple Linear Regression/mlr_rm.R
# install.packages('caTools', repos = "http://cran.us.r-project.org")
# install.packages('ggplot2', repos = "http://cran.us.r-project.org")


backwardsElimination <- function(x, sl) {
  numVars = length(x)
  for (i in c(1:numVars)){
    regressor = lm(formula = Profit ~., data=x)
    maxVar = max(coef(summary(regressor))[c(2:numVars), "Pr(>|t|)"])
    if (maxVar > sl){
      j = which(coef(summary(regressor))[c(2:numVars), "Pr(>|t|)"] == maxVar)
      x = x[,-j]
    }
    numVars = numVars - 1
  }
  return(regressor)
}


dataset = read.csv('/Users/rmaffesoli/Documents/personal/ML_A_Z/Machine Learning A-Z Template Folder/Part 2 - Regression/Section 5 - Multiple Linear Regression/50_Startups.csv')

# library(ggplot2)
library(caTools)

dataset$State = factor(dataset$State,
                       levels=c('New York', 'California', 'Florida'),
                       labels=c(1, 2, 3))


set.seed(123)
split = sample.split(dataset$Profit, SplitRatio = 0.8)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)
SL = 0.05
dataset = dataset[,c(1,2,3,4,5)]
regressor = backwardsElimination(training_set, SL)
print(summary(regressor))
y_pred = predict(regressor, test_set)

# regressor = lm(formula = Profit ~ .,
               # data=training_set)
# print(summary(regressor))
# regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State,
               # data=dataset)
# print(summary(regressor))
# regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend,
               # data=dataset)
# print(summary(regressor))
# regressor = lm(formula = Profit ~ R.D.Spend + Marketing.Spend,
               # data=dataset)
# print(summary(regressor))
# regressor = lm(formula = Profit ~ R.D.Spend,
               # data=dataset)
# print(summary(regressor))
